### PR TITLE
add maximum limit for grid of reduce, elementwise, gather and scatter

### PR DESCRIPTION
**Pull request by FlyingQianMM**
### PR types
Bug fixes

### PR changes
OPs

### Describe

The grid number of reduce、elementwise and masked_select has not been limited, which may raise a bug like:

```
:parallel_for failed: cudaErrorInvalidConfiguration: invalid configuration argument. 
```

So we add a maximum limit for grid of reduce, elementwise, gather and scatter kernel.

### CODE REVIEW

File: paddle/fluid/platform/device/gpu/gpu_launch_config.h [hunk 1]
```diff
@@ -128,6 +128,8 @@ inline GpuLaunchConfig GetGpuLaunchConfig1D(
   // Number of threads per block shall be larger than 64.
   threads = std::max(64, threads);
   int blocks = DivUp(DivUp(numel, vec_size), threads);
+  int limit_blocks = context.GetCUDAMaxGridDimSize()[0];
+  if (blocks > limit_blocks) blocks = limit_blocks;
```
Comment (ZzSean) [inline paddle/fluid/platform/device/gpu/gpu_launch_config.h]: 根据C++ Style要求，if条件最好加上大括号吧

File: paddle/phi/kernels/funcs/reduce_function.h [hunk 2]
```diff
@@ -1044,7 +1056,7 @@ void ReduceKernel(const KPDevice& dev_ctx,
 
   auto x_dim = phi::vectorize<int>(x.dims());
   auto config = ReduceConfig<Ty>(origin_reduce_dims, x_dim);
-  config.Run();
+  config.Run(x.place());
```
Comment (ZzSean) [inline paddle/phi/kernels/funcs/reduce_function.h:1059]: 可不可以把这个`LimitGridDim`写在外部，就可以直接用dev_ctx了

File: paddle/fluid/platform/device/gpu/gpu_launch_config.h (hunk 1 above)
Comment (FlyingQianMM) [reply paddle/fluid/platform/device/gpu/gpu_launch_config.h]: 已加上{}，感谢～

File: paddle/phi/kernels/funcs/reduce_function.h (hunk 2 above)
Comment (FlyingQianMM) [reply paddle/phi/kernels/funcs/reduce_function.h:1059]: config.Run()里面有对block数量做限制，所以把thread数量限制一起放在config.Run()里面了，这样可被复用

**Review by ZzSean — APPROVED** [review]
LGTM

**Comment by Xreki** [comment]
能不能给一下出错的算子配置、修改前的线程数和修改后的线程数？

**Pull request merged**